### PR TITLE
Visualise changes over time

### LINKED ISSUE #141: Consider how the visualiser handles a beneficial ownership network changing over a period of time
### Current situation
At the moment, all data is rendered according to the [specification here](https://github.com/openownership/visualisation-tool/blob/master/docs/spec.md).

No account is taken of fields like `Interest.endDate`, `replacesStatements` or field values like `Interest.share.exact`: 0 which indicate that information has changed over time, or that an interest is NOT present at a given time.

This means that when data represents change over time, it will either not be parsed ([see this OO Register/visualiser issue](https://github.com/openownership/register/issues/206)) or it may be visualised in a confusing way: it may look like a snapshot in time, but actually historical information is layered over current information.

With BODS 0.4 there will be clearer requirements regarding how to represent changing beneficial ownership over time. 

### What to do

We need to consider how the visualiser should handle data that represents change over time, both in BODS 0.4 and prior versions of the standard.

The minimal requirements is probably: have the visualiser explicitly represent information at a given point of time.

### LINKED ISSUE #142: Errors on BODS JSON from register
Using the [BODS JSON](https://register.openownership.org/entities/17979576658742990570.json) from https://register.openownership.org/entities/17979576658742990570 and the visualizer errors.

image

[17979576658742990570.json](https://github.com/openownership/visualisation-tool/files/13907776/17979576658742990570.json)

### LINKED ISSUE #174: Misaligned arrowheads
In working on #141, I've encountered a weird bug with the arrowheads:
Image

I believe it's because of the Bezier curves from [bezier.js](https://pomax.github.io/bezierjs/), so disabling the library entirely and using CSS transform to offset the lines, it improves, but the arrowheads are still misaligned:
Image

Two points:
1. We're currently using the `SVGtoBeziers()` function which is [not recommended to be used](https://github.com/Pomax/bezierjs/issues/129#issuecomment-674434665) and has since [been removed](https://github.com/Pomax/bezierjs/commit/4244adb73b42dfa90004e8986249eb0b1c84c9ea).
2. I don't think the CSS approach (second image) is very robust, and may break with larger graphs.

### LINKED ISSUE #176: Listed company icon should show for listed companies
On the dev branch at the moment, pasting in this BODS 0.3 data:

```
[
  {
    "entityType": "legalEntity",
    "foundingDate": "1992-06-17T00:00:00+00:00",
    "identifiers": [
      {
        "id": "02723534",
        "scheme": "GB-COH"
      },
      {
        "id": "PY6ZZQWO2IZFZC3IOL08",
        "scheme": "XI-LEI"
      }
    ],
    "jurisdiction": {
      "code": "GB",
      "name": "United Kingdom of Great Britain and Northern Ireland (the)"
    },
    "isComponent": false,
    "name": "AstraZeneca PLC",
    "publicListing": {
      "hasPublicListing": true,
      "companyFilingsURLs": [
        "https://www.londonstockexchange.com/stock/AZN/astrazeneca-plc/analysis",
        "https://www.astrazeneca.com/investor-relations/stock-exchange-announcements.html"
      ],
      "securitiesListings": [
        {
          "marketIdentifierCode": "XLON",
          "operatingMarketIdentifierCode": "XLON",
          "security": {
            "idScheme": "isin",
            "id": "GB0009895292",
            "ticker": "AZN"
          },
          "stockExchangeJurisdiction": "GB",
          "stockExchangeName": "London Stock Exchange"
        },
        {
          "marketIdentifierCode": "XLOM",
          "operatingMarketIdentifierCode": "XLON",
          "security": {
            "idScheme": "isin",
            "id": "US0463531089",
            "ticker": "0A4J"
          },
          "stockExchangeJurisdiction": "GB",
          "stockExchangeName": "London Stock Exchange - MTF"
        },
        {
          "marketIdentifierCode": "XSTO",
          "operatingMarketIdentifierCode": "XSTO",
          "security": {
            "idScheme": "isin",
            "id": "GB0009895292",
            "ticker": "AZN"
          },
          "stockExchangeJurisdiction": "SE",
          "stockExchangeName": "NASDAQ Stockholm"
        },
        {
          "marketIdentifierCode": "XNAS",
          "operatingMarketIdentifierCode": "XNAS",
          "security": {
            "idScheme": "cusip",
            "id": "046353108",
            "ticker": "AZN"
          },
          "stockExchangeJurisdiction": "US",
          "stockExchangeName": "NASDAQ"
        }
      ]
    },
    "publicationDetails": {
      "bodsVersion": "0.3",
      "license": "https://creativecommons.org/licenses/by/2.0/legalcode",
      "publicationDate": "2021-02-09",
      "publisher": {
        "name": "Open Ownership",
        "url": "http://www.openownership.org"
      }
    },
    "source": {
      "description": "This information has been retrieved from diverse public sources.",
      "type": [
        "primaryResearch"
      ]
    },
    "statementDate": "2021-02-09",
    "statementID": "openo-es-00006565580581270044236",
    "statementType": "entityStatement"
  }
]

```

erroneously shows as:

Image

Instead of:

Image

There is also a [BODS 0.4 example](https://raw.githubusercontent.com/openownership/data-standard/refs/heads/main/examples/listed-company-exempt-from-disclosure.json) that can be used to test a fix.

### LINKED ISSUE #177: Show error message when no visuals are renderable
Buried in this issue #70, was a requirement that:

When data is a valid JSON array of objects, but - after parsing and processing - actually no nodes or edges are rendered. It would be good to give user feedback in that instance. A simple error message like this would be useful:

> Your data does not have any information that can be drawn.

### Test data

```
[
  {
    "name": "nothing"
  }
]
```

### LINKED ISSUE #178: Redraw issue with properties tooltip and disclosure widget
### Bug description

When the content of the JSON text field is changed and 'Draw' pressed, if a properties tooltip is open, it remains open (along with the disclosure widget).

### Steps to reproduce

- Paste in this [JSON](https://raw.githubusercontent.com/openownership/data-standard/refs/heads/example_drt_main/examples/fermcat.json)
- Click on any edge or node so that the property tooltip is showing and the disclosure widget is populated.
- Delete the final ']' in the JSON text field (or otherwise change the text field content) and click 'Draw' 

### Ideal behaviour

The tooltip and disclosure widget would disappear as soon as 'Draw' is pressed.

### LINKED ISSUE #179: Page jumps when moving time slider
### Bug description

When the visualisation refreshes due to the time slider being dragged, if all elements of the tool are not already visible on the page, the page can appear to jump. 

### Steps to reproduce

1. Paste [this JSON](https://raw.githubusercontent.com/openownership/data-standard/refs/heads/example_drt_main/examples/fermcat.json) into the tool. Click 'Draw'.
2. Adjust zoom on the page so that only the diagram box and slider are visible on the page. E.g.

Image

3. Move the slider to a different position

### Expected behaviour

The page appears stable and only the diagram is updated.

### LINKED ISSUE #180: Operation of time slider handle
### Bug description

When the handle of the time slider is dragged to a new position, the user cannot drag it on to a further position without letting the handle go and then re-grabbing it.

### Steps to reproduce

1. Paste [this JSON](https://raw.githubusercontent.com/openownership/data-standard/refs/heads/example_drt_main/examples/fermcat.json) into the tool and click 'Draw'.
2. Click and hold down your mouse/pointing device to grab the time slider handle.
3. Drag the slider handle one step to the left, and maintain the grab action on the handle (keeping mouse button or equivalent pressed). 
4. Try to drag the handle one more step to the left. (You won't be able to.)
5. To drag the handle one more step to the left, release the handle and grab it again before moving it.

### Expected behaviour

Ideally the user would be able to drag the handle through multiple points on the time slider with a down-click -> move --> release action on their pointing device. (The diagram refreshing along the way.)

If this is too challenging to implement then some indication that the slider handle is no long under control would help the user.

### LINKED ISSUE #181: Mouse/pointer cursor on time slider handle could be more helpful to users
### Bug description

At the moment the standard pointer cursor is maintained throughout all interactions with the time slider handle. It would help users if clues were given for its operation.

### Steps to reproduce

1. Paste [this JSON](https://raw.githubusercontent.com/openownership/data-standard/refs/heads/example_drt_main/examples/fermcat.json) into the tool.
2. Drag your pointer to the time slider and along its length.

Image

3. Click and drag the handle.

### Expected behaviour

On mousing over the slide handle, change cursor to pointing hand (as happens on clicking a node or edge in the main diagram).

On clicking and holding the slide handle, change cursor to grab hand.

On releasing the slide handle, cursor is pointing hand.

On mousing over the line element of the slider, the cursor is the standard arrow pointer and the slider is highlighted (as now.)

### LINKED ISSUE #182: Properties tooltip and time slider behaviour
### Bug description

When the time point changes, if a properties tooltip is open, it remains open and does not update.

### Steps to reproduce

1. Paste [this JSON](https://raw.githubusercontent.com/openownership/data-standard/refs/heads/example_drt_main/examples/fermcat.json) into the visualiser tool and click 'Draw'.
2. Drag the timeslider handle to the 2020-09-11 point.
3. Click the 'Patrick O'Donohue' node. A tooltip pops up and the disclosure widget is populated and open.
4. Drag the timeslider handle one step to the right, to the 2021-09-11 point.
5. Notice that the disclosure widget is closed, but the tooltip remains in position and contains the content of the 'old' node.

### Expected behaviour

Like the disclosure widget, the tooltip should disappear once the handle on the timeslider moves.

### LINKED ISSUE #184: Too many fields required to render a node
### Bug description

In the spec we say that the minimum fields required to render an entity node are:

- `recordType` is “entity”
- if `recordStatus` exists, it is not “closed”

And the equivalent for person nodes.

But in the demo, more fields than this have to be included.

### Steps to reproduce

1.  Paste the following into the tool and click 'Draw'

```
[
  {
    "statementDate": "2019-01-20",
    "recordType": "entity",
    "recordDetails": {
      "entityType": {}
    }
  }
]
```
2. Notice that an unknown entity node is rendered.

3. Remove any one of the fields and try to redraw .

For person nodes, this is currently the minimal json that renders a node:

```
[
  {
    "statementDate": "2019-01-20",
    "recordType": "person"
  }
]
```

### Expected behaviour

Ideally the following should render a node:

```
[
  {
    "recordType": "entity"
  }
]
```

And the equivalent for "person".

If making that work causes too many knock-on problems at this stage, we can change the spec.

### LINKED ISSUE #185: Improve experience of recovering from a JSON error
### Description of current behaviour

When the data submitted to the tool isn't valid JSON, a dialog box warning appears saying 'There was an error drawing your data. The data must be a valid JSON array of objects.'. 

After 'OK' is clicked and the box is closed, both the diagram box and the JSON text area are cleared, and 'undefined' appears in the text area. This gives no opportunity to the user to review the contents of the text area and correct it.

### Steps to reproduce

1. Paste the following into the tool and click 'Draw'

```
[
  {
    "statementDate": "2019-01-20",
    "recordId": "01B68D7633",
    "recordType": "entity",
    "recordDetails": {
      "entityType": {
        "type": ""
      }
    }
  },
  {
    "statementDate": "2019-01-20",
    "recordId": "01B68D7633",
    "recordType": "relationship",
    "recordDetails": {
      "interestedParty": "31B68D7633",
      "subject": "01B68D7633"
    }
  },
  {
    "statementDate": "2019-01-20",
    "recordId": "31B68D7633",
    "recordType": "person"
  }
]
```
You'll see a minimal diagram rendered.

2. Now add the following property to the first statement BUT intentionally drop a comma so that it invalidates the JSON

```
"recordStatus": "new"
```

3. Click 'Draw' again

### Suggested behaviour

After clicking 'OK' on the error message, the diagram box should be cleared, but the contents of the JSON text area should remain unchanged.

### PR DESCRIPTION
**Pull request by codemacabre**
- Closes #141
- Closes #142
- Closes #174
- Closes #176
- Closes #177
- Closes #178
- Closes #179
- Closes #180
- Closes #181
- Closes #182
- Closes #184
- Closes #185

~~Note, this update introduces a bug with the arrowheads which will be addressed in a future issue (see #174).~~

### CODE REVIEW

File: src/utils/bods.js [hunk 1]
```diff
@@ -1,5 +1,42 @@
...
+    }
+
+    const duplicates = statements.filter((statement) => {
+      const key = `${statement.recordId}-${statement.recordType}`;
+      return recordIdCount[key] > 1;
+    });
+
+    console.log(duplicates);
+  } else {
+    // get all statements with statementID values in replacesStatements array
```
Comment (radix0000) [inline src/utils/bods.js]: What is the deal with this? Does this PR add rendering change over time just for BODS 0.4? If so probably best to be explicit about that in this comment?

**Comment by radix0000** [comment]
Okay, had to hand roll my own BODS 0.3 test data (because examples are so broken):

https://gist.github.com/radix0000/5e612d582b41a8aca5bfff4d5e396bf8

So doesn't show change over time with BODS 0.3 (which is fine if that is the intention), but does go into "change-over-time mode" with slider displayed if there is a replacesStatement property present (you just can't slide the slider, and all statements displayed), but latest time it shows conflicts with the fact all statements are shown, which is a bit confusing. So if it is intended that doesn't support change-over-time for 0.1-0.3 then recon that slider should always be hidden regardless of whether there is a replacesStatement property or not.

File: src/utils/bods.js (hunk 1 above)
Comment (codemacabre) [reply src/utils/bods.js]: > Does this PR add rendering change over time just for BODS 0.4?

Yes, the [specification](https://docs.google.com/document/d/15YBQbPixnS1i1qe4RlXfkx8wchVsvuEI3ZrCS_pOg5A/edit?tab=t.0) states that only 0.4 data renders an interactive slider. I'll update the comment (and remove the `console.log`).

**Comment by codemacabre** [comment]
@radix0000 Thank you. @kd-ods and I discussed this at length and [came to the decision](https://github.com/openownership/visualisation-tool/issues/141#issuecomment-2470415956) that 0.4 data renders and interactive slider, whereas pre-0.4 data with a change-over-time element renders an inactive slider to indicate that there is change-over-time data, but not render the changes themselves. The slider is hidden only if there is no change-over-time data detected.

**Review by radix0000 — APPROVED** [review]
Looks good. One thing I happened to notice when testing is that data with the wrong version (i.e. bodsVersion == "0.4" when it is actually 0.3 format) results in a "Your data does not have any information that can be drawn." message, and while it should certainly give an error, ideally it could be more informative (especially as the only difference between data that works and doesn't, is one character). Possibly having a check for "recordDetails" property and whether its presence or not matches the version, < 0.4 vs >= 0.4, could work. I am doing something similar in the data review tool). But don't think it is a show stopper, so happy for this to be merged either way.

**Pull request merged**